code/lora/lora.py

```diff
@@ -0,0 +1,298 @@
+        and transmits it.
+        """
+
+        if len(self.buffer) == 0:
+            print("Buffer empty!")
+            self.buffer.insert(0, "SSBuffer Empty!")
+
+        payload_str = self.buffer.pop(0)
+        payload_str += "-"*(DOWN_PACKET_LENGTH-len(payload_str))
+
+        payload_b = list(bytearray(payload_str, "utf-8"))
+        payload = HEADER.LORA + payload_b
+        print("TX: " + payload_str)
+
+        self.write_payload(payload)
+        self.set_mode(MODE.TX)
+        time.sleep(TX_TIME)
+        self.reset_ptr_rx()
+        self.set_mode(MODE.RXCONT)
+        self.last_tx_mode = 'default'
+
+
+    def tx_image(self):
+        """
+        Transmits a chunk of picamera image data.
+        If no ACK was received, same payload is sent.
+        Otherwise a new chunk is loaded.
+        Upon reaching end of file, stops the transmission.
+        """
+
+        if self.ack['image']: #If ACK received, send next slice of data
+            self.image_waiting = [HEADER.IMAGE_B] + list(self.image.read(DOWN_PACKET_LENGTH-1))
+            if self.image_waiting == [HEADER.IMAGE_B]: #If EOF, end image transmission
+                self.sending_image = False
+                return
+
+        self.ack['image'] = False
+        packet_length = len(self.image_waiting)
+        if packet_length < DOWN_PACKET_LENGTH:
+            self.image_waiting += (DOWN_PACKET_LENGTH-packet_length)*[45]
+        print("TX:", self.image_waiting)
+
+        self.write_payload(HEADER.LORA + self.image_waiting)
+        self.set_mode(MODE.TX)
+        time.sleep(TX_TIME)
+        self.reset_ptr_rx()
+        self.set_mode(MODE.RXCONT)
+        self.last_tx_mode = 'image'
+
+
+    def tx_thermal(self):
+        """
+        Transmits a chunk of MLX90640 thermal image.
+        If no ACK was received, same payload is sent.
+        Otherwise, a new chunk is loaded.
+        Upon transmitting 32 lines, stops the transmission
+        """
+        #NOT TESTED YET
+
+        if self.ack['thermal']:
+            self.thermal_counter += 1
+            if self.thermal_counter > THERMAL_HEIGHT:
+                self.sending_thermal = False
+                return
+            row = [int(i) for i in self.thermal[self.thermal_counter-1]]
+            row[:] = [max(el, 0) if el < 255 else min(el, 255) for el in row]
+            self.thermal_waiting = [HEADER.THERMAL_B] + row
+
+        print(self.thermal_waiting)
+        self.ack['thermal'] = False
+        self.write_payload(HEADER.LORA + self.thermal_waiting)
+        self.set_mode(MODE.TX)
+        time.sleep(TX_TIME)
+        self.reset_ptr_rx()
+        self.set_mode(MODE.RXCONT)
+        self.last_tx_mode = 'thermal'
+
+
+    def wait_for_rx(self):
+        """
+        Waits for slave responce.
+        No sleep() because of expected RX callback
+        """
+
+        start_time = time.time()
+        while time.time() - start_time < RX_TIME:
+            pass
```

Comment:
small sleep is needed or we're gonna burn the cpu to death. I mean, this `while loop` will take up to 100% of CPU and won't leave any resources for other operating system tasks and so on

Reply:
Oh. 
Such solution was implemented in the sample code from the lib. 
If we use sleep(), will the on_rx_done callback work as intended?

Reply:
I peeked into the `pySX127x` module and I don't see any callbacks going on in there. Correct me if I'm wrong but you should just read from `read_payload` method

Reply:
Well, I also dove in there, and there is in fact a GPIO callback. See line 101 in board_config.py